[Bifrost] Implements seal() for LocalLoglet

AhmedSoliman opened the pull request:
[Bifrost] Implements seal() for LocalLoglet

---
[//]: # (BEGIN SAPLING FOOTER)
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/restatedev/restate/pull/1718).
* #1726
* #1725
* #1722
* #1720
* __->__ #1718
* #1717
* #1705
* #1704

tillrohrmann commented on crates/bifrost/src/providers/local_loglet/mod.rs:
```diff
@@ -230,14 +234,31 @@ impl LogletBase for LocalLoglet {
+                .max(offset.into()),
+        );
+        self.notify_readers(release_pointer);
+        // Ensure that we don't acknowledge the append (even that it has happened) if the loglet
+        // has been sealed already. Memory ordering Acquire is required to establish happens-before
+        // relationship between bumping the last_commit_offset and the seal operation.
+        if self.sealed.load(Ordering::Acquire) {
```
I think I don't fully understand the reasoning here. If we read `sealed == true`, it makes sense to abort the append. Couldn't this read be done with `Ordering::Relaxed` since we don't depend on any other memory writes that happened before setting `sealed = true`? If we read `sealed == false`, this is probably the dangerous situation since we don't know whether the write (self.sealed = true) is still in the store buffer and we have already read `self.last_committed_offset` in `find_tail` before we incremented it in this method?

tillrohrmann reviewed (commented):
Thanks a lot for implementing seal for the `LocalLoglet`! Whenever I have to deal with memory orderings, I am on thin ice. So maybe the comments I've left don't make sense at all. My main question was to clarify my understanding about why certain accesses are `SeqCst` and `Acquire`/`Release` and how the problematic scenario of someone having sealed the loglet and reported a tail and a concurrent append operation being prevented from succeeding works.

AhmedSoliman commented on crates/bifrost/src/providers/local_loglet/mod.rs:
```diff
@@ -491,4 +534,87 @@ mod tests {
+                        log_writer,
+                    )
+                    .await?,
+                );
```
I can do that in a future follow up PR

AhmedSoliman commented on crates/bifrost/src/providers/local_loglet/mod.rs:
```diff
@@ -230,14 +234,31 @@ impl LogletBase for LocalLoglet {
+        // relationship between bumping the last_commit_offset and the seal operation.
+        if self.sealed.load(Ordering::Acquire) {
+            return Err(AppendError::Sealed);
+        }
```
I really like the idea of fetch_add(0, Release), very clever solution! 

In all fairness, my approach was to mash something together and slam SeqCst as temporary measure until a follow up PR that will replace the entire module with a sequencer implementation. But since you've done a great job reviewing it, I've updated this version to a nicer/better version. 

In fact, what was making the whole thing work is the Mutex that already acts as a memory barrier/fence which synchronized threads.

AhmedSoliman commented on crates/bifrost/src/providers/local_loglet/mod.rs:
```diff
@@ -295,7 +329,7 @@ impl LogletBase for LocalLoglet {
     async fn trim(&self, new_trim_point: Self::Offset) -> Result<(), OperationError> {
         let effective_trim_point = new_trim_point.min(LogletOffset(
-            self.last_committed_offset.load(Ordering::Relaxed),
+            self.last_committed_offset.load(Ordering::Acquire),
```
Not needed

AhmedSoliman commented on crates/bifrost/src/providers/local_loglet/mod.rs:
```diff
@@ -329,7 +363,16 @@ impl LogletBase for LocalLoglet {
+            warn!("Unsure if the local loglet record was sealed, the ack channel was dropped");
+            Err(ShutdownError.into())
+        })?;
+        self.sealed.store(true, Ordering::Release);
```
Not needed anymore.

AhmedSoliman commented:
I've cleaned up and fixed this implementation, improved the test to tackle the corner case a little harder (find_tail during appends + seal), and validated the approach in an external loom test (albeit, loom tests can't catch all issues with Relaxed ordering, so I opted to use `AcqRel` in fetch_max which arguably can be relaxed more.

tillrohrmann reviewed (approved):
Great work @AhmedSoliman. As far as I can tell, the atomic operations seem sound to me. +1 for merging.

I've never used Loom so I'd be interested in taking a look at the test and to run it to see how it works :-)
